portal-api/src/v1/controllers/utilisation-report-service/previous-reports.controller.ts

```diff
@@ -0,0 +1,89 @@
+import { AxiosError } from 'axios';
+import { Request, Response } from 'express';
+import orderBy from 'lodash/orderBy';
+import api from '../../api';
+import { UtilisationReportResponseBody } from '../../api-response-types';
+
+/**
+ * Returns a set of unique years based on reports from the database.
+ * @param reports - reports from the database
+ * @returns unique set of years
+ */
+export const getYears = (reports: UtilisationReportResponseBody[]): number[] => {
+  const years = reports.map((report) => report.reportPeriod.start.year);
+  return [...new Set(years)];
+};
+
+type GroupedReport = {
+  year: number;
+  reports: UtilisationReportResponseBody[];
+};
+
+/**
+ * Groups database reports by year
+ * @param years - unique set of years
+ * @param reports - reports from the database
+ * @returns - list of objects with year and reports property
+ */
+export const groupReportsByStartYear = (years: number[], reports: UtilisationReportResponseBody[]): GroupedReport[] =>
+  years.map((year) => ({
+    year,
+    reports: reports.filter((report) => report.reportPeriod.start.year === year),
+  }));
+
+/**
+ * Adds an object for all year with is no database reports when there is a report for the
+ * previous and future years
+ * @param reportsGroupedByStartYear - list of objects with year and reports property
+ * @param years - unique set of years
+ * @returns list of objects with year and reports property
+ */
+export const populateOmittedYears = (reportsGroupedByStartYear: GroupedReport[], years: number[]): GroupedReport[] => {
+  years.forEach((year, index) => {
+    if (index > 0 && year - years[index - 1] > 1) {
+      let numberOfMissingYears = year - years[index - 1] - 1;
+      while (numberOfMissingYears > 0) {
+        reportsGroupedByStartYear.push({
+          year: year - numberOfMissingYears,
+          reports: [],
+        });
+        numberOfMissingYears -= 1;
+      }
+    }
+  });
+  return reportsGroupedByStartYear;
+};
+
+/**
+ * Groups the reports by year and sorts year descending
+ * @param dbReports - reports from the database
+ * @returns list of objects with year and reports property
+ */
+export const groupAndSortReports = (dbReports: UtilisationReportResponseBody[]): GroupedReport[] => {
+  const years = getYears(dbReports);
+  const groupedReports = groupReportsByStartYear(years, dbReports);
+  const groupedReportsByStartYear = populateOmittedYears(groupedReports, years);
+  return orderBy(groupedReportsByStartYear, ['year'], ['desc']);
+};
+
+type GetPreviousReportsByBankIdRequest = Request<{ bankId: string }>;
+
+export const getPreviousReportsByBankId = async (req: GetPreviousReportsByBankIdRequest, res: Response) => {
```

Comment:
The return type of this controller has always been a pain for me to figure out each time I have to look at it, so I took this as an opportunity to convert this file to typescript and document what gets returned